### PR TITLE
Do not warn user of incompatible SDK version if they are on a preview

### PR DESCRIPTION
**Pull request by jmarolf**
fixes internal bug
- [819411](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/819411)

This change now makes it so there is an additional `preview` version that is taken into account (if the user is on a preview release of VS or has chosen to use a preview SDK).

built on top of 
- https://github.com/dotnet/project-system/pull/4712
- https://github.com/dotnet/project-system/pull/4711

review commits https://github.com/dotnet/project-system/commit/527a14969ae39df5adc08c99fc534547b3f7188d to https://github.com/dotnet/project-system/commit/04a6c06bc983c1110d550ada980d1ac95255c44b

Open question: do we want different behavior for visual studio previews vs. the use of preview sdks?

### CODE REVIEW

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/DotNetCoreProjectCompatibilityDetector.cs [hunk 1]
```diff
@@ -335,35 +335,61 @@ private static CompatibilityLevel GetCompatibilityLevelFromVersion(Version versi
...
+            if (supportedPreviewVersion is object && IsPreview() && version <= supportedPreviewVersion)
+            {
+                return CompatibilityLevel.Recommended;
+            }
+
+            return GetCompatibilityLevelFromSupportedn(version, supportedVersion, unsupportedVersion);
+        }
 
-                return CompatibilityLevel.NotSupported;
+        private static CompatibilityLevel GetCompatibilityLevelFromSupportedn(Version version, Version supportedVersion, Version unsupportedVersion)
```
Comment (Pilchie) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/DotNetCoreProjectCompatibilityDetector.cs]: Typo: Supportedn

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/DotNetCoreProjectCompatibilityDetector.cs [hunk 2]
```diff
@@ -335,35 +335,61 @@ private static CompatibilityLevel GetCompatibilityLevelFromVersion(Version versi
...
+        {
+            // a supported version exists and the version is less than the supported version
+            if (supportedVersion is object && version < supportedVersion)
+            {
+                return CompatibilityLevel.Recommended;
             }
 
-            // Only has an unsupported version
-            if (compatData.UnsupportedVersion != null)
+            // ther version is not unsupported and exactly matches the supported version
```
Comment (Pilchie) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/DotNetCoreProjectCompatibilityDetector.cs]: Spelling: "ther"

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/DotNetCoreProjectCompatibilityDetector.cs [hunk 3]
```diff
@@ -335,35 +335,61 @@ private static CompatibilityLevel GetCompatibilityLevelFromVersion(Version versi
...
             {
-                if (version < compatData.SupportedVersion)
-                {
-                    return CompatibilityLevel.Recommended;
-                }
-                else if (version == compatData.SupportedVersion || (compatData.UnsupportedVersion != null && version < compatData.UnsupportedVersion))
-                {
-                    return CompatibilityLevel.Supported;
-                }
+                // no restrictions
```
Comment (Pilchie) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/DotNetCoreProjectCompatibilityDetector.cs]: Nit: throught - normally we capitalize the first letter of comments (applies throughout)

**Comment by Pilchie** [comment]
For the open question, I think what we wanted was to be able to have a message only on Release channel of VS to say to check the checkbox, even though the VS Version and SDK Version might be identical.

**Comment by jmarolf** [comment]
> 
> 
> For the open question, I think what we wanted was to be able to have a message only on Release channel of VS to say to check the checkbox, even though the VS Version and SDK Version might be identical.

@Pilchie this should be able to do that.  You provide an Preview SDK version that is supported and on preview releases of VS / VS RTM *with* the checkbox checked we do not show the user a message.  In VS RTM we will

**Pull request merged**